examples/step-50/step-50.cc

```diff
@@ -14,912 +14,1427 @@
+  template <typename number>
+  number average_value(const std::vector<Point<dim, number>> &points) const;
+};
 
-    AffineConstraints<double> constraints;
 
-    VectorType solution;
-    VectorType system_rhs;
+template <int dim>
+double Coefficient<dim>::value(const Point<dim> &p, const unsigned int) const
+{
+  for (int d = 0; d < dim; ++d)
+    {
+      if (p[d] < -0.5)
+        return 100.0;
+    }
+  return 1.0;
+}
 
-    const unsigned int degree;
 
-    // Finally we are storing the various parallel multigrid matrices. Our
-    // problem is self-adjoint, so the interface matrices are the transpose
-    // of each other, so we only need to compute/store them once.
-    MGLevelObject<MatrixType> mg_matrices;
-    MGLevelObject<MatrixType> mg_interface_matrices;
-    //
-    MGConstrainedDoFs mg_constrained_dofs;
-  };
+template <int dim>
+template <typename number>
+VectorizedArray<number>
+Coefficient<dim>::value(const Point<dim, VectorizedArray<number>> &p,
+                        const unsigned int) const
+{
+  VectorizedArray<number> return_value = VectorizedArray<number>();
+  for (unsigned int i = 0; i < VectorizedArray<number>::size(); ++i)
+    {
+      bool found = false;
+      for (int d = 0; d < dim; ++d)
+        if (p[d][i] < -0.5)
+          {
+            return_value[i] = 100.0;
+            found           = true;
+            break;
+          }
+
+      if (!found)
+        return_value[i] = 1.0;
+    }
+
+  return return_value;
+}
 
 
+template <int dim>
+template <typename number>
+number Coefficient<dim>::average_value(
+  const std::vector<Point<dim, number>> &points) const
+{
+  number average(0);
+  for (unsigned int i = 0; i < points.size(); ++i)
+    average += value(points[i]);
+  average /= points.size();
 
-  // @sect3{Nonconstant coefficients}
+  return average;
+}
 
-  // The implementation of nonconstant
-  // coefficients is copied verbatim
-  // from step-5 and step-6:
 
-  template <int dim>
-  class Coefficient : public Function<dim>
-  {
-  public:
-    virtual double value(const Point<dim> & p,
-                         const unsigned int component = 0) const override;
 
-    virtual void value_list(const std::vector<Point<dim>> &points,
-                            std::vector<double> &          values,
-                            const unsigned int component = 0) const override;
-  };
+/**
+ * Matrix-free Laplace operator
+ */
+template <int dim, int fe_degree, typename number>
+class LaplaceOperator
+  : public MatrixFreeOperators::Base<dim,
+                                     LinearAlgebra::distributed::Vector<number>>
+{
+  using VectorizedArrayType = VectorizedArray<number>;
+
+public:
+  LaplaceOperator();
 
+  void clear() override;
 
+  void evaluate_coefficient(const Coefficient<dim> &coefficient_function);
 
-  template <int dim>
-  double Coefficient<dim>::value(const Point<dim> &p, const unsigned int) const
+  AlignedVector<VectorizedArrayType> get_coefficient_vector()
```

Comment:
Do you really want to return a copy of the coefficient rather than a const reference?
```suggestion
  const AlignedVector<VectorizedArrayType> & get_coefficient_vector() const
```